docs/index.bs

```diff
@@ -1545,6 +1554,49 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
     Note: [=/Service workers=] delay treating the [=installing worker=] as "`installed`" until all the [=promises=] in the {{install!!event}} event's [=extend lifetime promises=] resolve successfully. (See the relevant [Install algorithm step](#install-settle-step).) If any of the promises rejects, the installation fails. This is primarily used to ensure that a [=/service worker=] is not considered "`installed`" until all of the core caches it depends on are populated. Likewise, [=/service workers=] delay treating the [=active worker=] as "`activated`" until all the [=promises=] in the {{activate!!event}} event's [=extend lifetime promises=] settle. (See the relevant [Activate algorithm step](#activate-settle-step).) This is primarily used to ensure that any [=functional events=] are not dispatched to the [=/service worker=] until it upgrades database schemas and deletes the outdated cache entries.
   </section>
 
+  <section>
+    <h3 id="installevent-interface">{{InstallEvent}}</h3>
+
+    <pre class="idl">
+      [Exposed=ServiceWorker]
+      interface InstallEvent : ExtendableEvent {
+        Promise&lt;undefined&gt; registerRouter((RouterRule or sequence&lt;RouterRule&gt;) rules);
+      };
+
+      dictionary RouterRule {
+        required RouterCondition condition;
+        required RouterSourceEnum source;
+      };
+
+      dictionary RouterCondition {
+        USVString urlPattern;
+      };
+
+      enum RouterSourceEnum { "network" };
+    </pre>
+
+    Each {{RouterCondition/urlPattern}} object has an associated <dfn>hostname component</dfn>, a [=urlpattern/component=], which is initially unset.
```

Comment:
This doesn't really work. You can't associate a component with a string inside a dictionary.

I think the way to do this is to create a parallel hierarchy of [structs](https://infra.spec.whatwg.org/#structs) mirroring your dictionaries, that represent the parsed result. E.g. a "router rule" struct with:

- source (a `RouterSourceEnum`)
- condition (a router condition struct)

and then a "router condition" struct with

- urlPattern, a component

Then the static router rules object should be a list of parsed "router rules", not a list of `RouterRule`s.

Reply:
I tried to follow what https://wicg.github.io/urlpattern/#urlpattern-pathname-component does, but I am doing that well or not.
The components should be an internal structure and it should not be directly accessed by web facing APIs.
Are there any examples having a hidden data structure?

Reply:
Anyway, now it is URLInput inside a directory, and the comment get outdated.  Please feel free to leave a new comment or reopen.